### PR TITLE
[bugfix] Prevent deepcopy of dataloaders / Trainer in SWA Callback

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces a context manager to prevent deecopying trainer / dataloaders from the LightningModule.
This PR uses this context manager to prevent with stochastic weight averaging
This PR improve ModelPruning save_checkpoint to avoid a deepcopy.

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/7089

### Does your PR introduce any breaking changes ? If yes, please list them.

No.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 1]
```diff
@@ -137,9 +138,32 @@ def swa_end(self) -> int:
...
+        pl_module.val_dataloader = None
+        pl_module.test_dataloader = None
+        pl_module.predict_dataloader = None
+        pl_module.trainer = None
+        yield
+        pl_module.train_dataloader = train_dataloader
+        pl_module.val_dataloader = val_dataloader
+        pl_module.test_dataloader = test_dataloader
+        pl_module.predict_dataloader = predict_dataloader
+        pl_module.trainer = trainer
```
Comment (carmocca) [inline pytorch_lightning/callbacks/stochastic_weight_avg.py:161]: We could do this by defining `__getstate__` on the `LightningModule` instead

Comment (tchaton) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py:161]: Not sure we want to, some users might expect a complete deepcopy. At least, it is contained to the SWA Callback for now.

Comment (tchaton) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py:161]: @justusschock @awaelchli  Any opinions ?

Comment (awaelchli) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py:161]: what else could be problematic for deepcopy in LightningModule? User might have some problematic objects?
Would they be able to get around it by implementing getstate?

Comment (awaelchli) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py:161]: I don't have any strong opinions who this should be handled

Comment (carmocca) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py:161]: I think this would also be beneficial for the `ModelPruning` callback which also deepcopies the model. This is why I pointed it out.

Comment (tchaton) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py]: Done !

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1998,3 +1998,10 @@ def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
         self.trainer.callback_metrics.update(
             apply_to_collection(callback_metrics, np.ndarray, lambda x: torch.tensor(x))
         )
+
+    def __getstate__(self) -> Dict[str, Any]:
+        # prevent copying data
+        skipped_keys = {
+            "trainer", "train_dataloader", "val_dataloader", "test_dataloader", "predict_dataloader", "datamodule"
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:2005]: Why do you skip the dataloader ones which are functions, not properties or attributes?

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 3]
```diff
@@ -217,3 +217,29 @@ def configure_optimizers(self):
...
+
+    class TestSWA(StochasticWeightAveraging):
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.on_before_accelerator_backend_setup_called = False
+
+        def on_before_accelerator_backend_setup(self, trainer: 'Trainer', pl_module: 'LightningModule'):
+            super().on_before_accelerator_backend_setup(trainer, pl_module)
+            assert self._average_model.train_dataloader is not pl_module.train_dataloader
```
Comment (carmocca) [inline tests/callbacks/test_stochastic_weight_avg.py:233]: What are you trying to test with this line?

**Review by carmocca — COMMENTED** [review]
Pushed a few commits with some changes

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/core/lightning.py:2005]: Just guessing, but if the user provides the loaders to the trainer, these functions are patched with callable classes also containing the loader object.

File: tests/callbacks/test_stochastic_weight_avg.py (hunk 3 above)
Comment (tchaton) [reply tests/callbacks/test_stochastic_weight_avg.py:233]: The pl_module train_dataloader has been wrapped into a PatchDataLoader at this stage. However, the train_dataloader on the average model should be the default function.

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py]: This doesn't play nice with ddp_spawn and ddp_cpu. I needed to make this optional through a context manager.

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/callbacks/pruning.py [hunk 4]
```diff
@@ -407,6 +408,23 @@ def on_train_end(self, trainer: 'pl.Trainer', pl_module: LightningModule) -> Non
             rank_zero_debug("`ModelPruning.on_train_end`. Pruning is made permanent for this checkpoint")
             self.make_pruning_permanent(pl_module)
 
+    def _make_state_dict_permanent(self, pl_module: LightningModule) -> Dict[str, Any]:
+        state_dict = pl_module.state_dict()
+
+        # find the mask and the original weights.
+        map_pruned_params = {k.replace("_mask", "") for k in state_dict.keys() if k.endswith("_mask")}
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:415]: Don't we have access to the list of parameters to prune with `self._parameters_to_prune`?

File: pytorch_lightning/callbacks/pruning.py [hunk 5]
```diff
@@ -407,6 +408,23 @@ def on_train_end(self, trainer: 'pl.Trainer', pl_module: LightningModule) -> Non
...
+        for tensor_name in map_pruned_params:
+            orig = state_dict.pop(tensor_name + "_orig")
+            mask = state_dict.pop(tensor_name + "_mask")
+            # make weights permanent
+            state_dict[tensor_name] = mask.to(dtype=orig.dtype) * orig
+
+        def move_to_cpu(tensor: torch.Tensor) -> torch.Tensor:
+            # deepcopy each tensor and move them on cpu
+            return deepcopy(tensor).cpu()
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:424]: Why do you need to `deepcopy`? `state_dict()` copies it already

File: pytorch_lightning/callbacks/pruning.py [hunk 6]
```diff
@@ -407,6 +408,23 @@ def on_train_end(self, trainer: 'pl.Trainer', pl_module: LightningModule) -> Non
...
+    def _make_state_dict_permanent(self, pl_module: LightningModule) -> Dict[str, Any]:
+        state_dict = pl_module.state_dict()
+
+        # find the mask and the original weights.
+        map_pruned_params = {k.replace("_mask", "") for k in state_dict.keys() if k.endswith("_mask")}
+        for tensor_name in map_pruned_params:
+            orig = state_dict.pop(tensor_name + "_orig")
+            mask = state_dict.pop(tensor_name + "_mask")
+            # make weights permanent
+            state_dict[tensor_name] = mask.to(dtype=orig.dtype) * orig
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:420]: Won't the mask be of the correct `dtype` already?

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1998,3 +2001,17 @@ def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
         self.trainer.callback_metrics.update(
             apply_to_collection(callback_metrics, np.ndarray, lambda x: torch.tensor(x))
         )
+
+    @contextmanager
+    def prevent_trainer_and_dataloaders_deepcopy(self) -> None:
+        self._prevent_trainer_and_dataloaders_deepcopy = True
+        yield
+        self._prevent_trainer_and_dataloaders_deepcopy = False
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:2009]: if this is a pubic method now, we would have to document it.
but doesn't it make more sense as a protected context manager?

File: pytorch_lightning/callbacks/pruning.py [hunk 8]
```diff
@@ -415,13 +433,8 @@ def on_save_checkpoint(
     ) -> Dict[str, Any]:
         if self._make_pruning_permanent:
             rank_zero_debug("`ModelPruning.on_save_checkpoint`. Pruning is made permanent for this checkpoint")
-            prev_device = pl_module.device
-            # prune a copy so training can continue with the same buffers
-            copy = deepcopy(pl_module.to("cpu"))
-            self.make_pruning_permanent(copy)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:421]: `make_pruning_permanent` handles the pruning forward hooks, not just the mask multiplication.

Are you sure not using it is a good idea? Don't you think we could be missing special logic by handling the pruning ourselves?

File: pytorch_lightning/callbacks/pruning.py [hunk 9]
```diff
@@ -407,6 +408,23 @@ def on_train_end(self, trainer: 'pl.Trainer', pl_module: LightningModule) -> Non
             rank_zero_debug("`ModelPruning.on_train_end`. Pruning is made permanent for this checkpoint")
             self.make_pruning_permanent(pl_module)
 
+    def _make_state_dict_permanent(self, pl_module: LightningModule) -> Dict[str, Any]:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:411]:
```suggestion
    def _make_pruning_permanent_on_state_dict(self, pl_module: LightningModule) -> Dict[str, Any]:
```

File: pytorch_lightning/core/lightning.py [hunk 10]
```diff
@@ -112,6 +113,8 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         # deprecated, will be removed in 1.6
         self._loaded_optimizer_states_dict = {}
 
+        self._prevent_trainer_and_dataloaders_deepcopy: bool = False
+
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:117]: Can you move this to L112 with the rest of non-deprecated attributes?

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 11]
```diff
@@ -217,3 +218,31 @@ def configure_optimizers(self):
...
+            self.on_before_accelerator_backend_setup_called = True
+
+    model = BoringModel()
+    swa = TestSWA()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        callbacks=swa,
+        fast_dev_run=True,
+    )
+    trainer.fit(model, train_dataloader=DataLoader(RandomDataset(32, 2)))
```
Comment (carmocca) [inline tests/callbacks/test_stochastic_weight_avg.py:247]: why do you need to pass a dataloader?

File: pytorch_lightning/core/lightning.py [hunk 12]
```diff
@@ -1998,3 +2001,17 @@ def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
...
+        yield
+        self._prevent_trainer_and_dataloaders_deepcopy = False
+
+    def __getstate__(self) -> Dict[str, Any]:
+        skip_keys = set()
+        if self._prevent_trainer_and_dataloaders_deepcopy:
+            skip_keys.update({
+                "trainer", "train_dataloader", "val_dataloader", "test_dataloader", "predict_dataloader", "datamodule"
+            })
+        return {k: v for k, v in self.__dict__.items() if k not in skip_keys}
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:2017]: I prefer the implementation where the skip_keys are set to None. It's weird when the property doesn't exist but it's defined in the source code

**Review by carmocca — CHANGES_REQUESTED** [review]

**Review by awaelchli — COMMENTED** [review]
If I understand this PR correctly, all the changes are specific to the pruning etc. callbacks. can you confirm the following:
- pickling behavior unchanged
- deepcopy(model) still does a real deepcopy unless the context manager is used
- the context manager is not meant to be used by the user, or is it? any other use cases for deepcopying a lightnignmodule?

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:2009]: Yes, let's make it private.

File: pytorch_lightning/callbacks/pruning.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/pruning.py:424]: Removed !

File: pytorch_lightning/callbacks/pruning.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/pruning.py:420]: I checked internally and that's how they compute it.

File: pytorch_lightning/callbacks/pruning.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/pruning.py:421]: Here is what the remove function does. This is not required as we keep training and we can just do it manually.

```py
 def remove(self, module):
        r"""Removes the pruning reparameterization from a module. The pruned
        parameter named ``name`` remains permanently pruned, and the parameter
        named ``name+'_orig'`` is removed from the parameter list. Similarly,
        the buffer named ``name+'_mask'`` is removed from the buffers.

        Note:
            Pruning itself is NOT undone or reversed!
        """
        # before removing pruning from a tensor, it has to have been applied
        assert (
            self._tensor_name is not None
        ), "Module {} has to be pruned\
            before pruning can be removed".format(
            module
        )  # this gets set in apply()

        # to update module[name] to latest trained weights
        weight = self.apply_mask(module)  # masked weights

        # delete and reset
        if hasattr(module, self._tensor_name):
            delattr(module, self._tensor_name)
        orig = module._parameters[self._tensor_name + "_orig"]
        orig.data = weight.data
        del module._parameters[self._tensor_name + "_orig"]
        del module._buffers[self._tensor_name + "_mask"]
        setattr(module, self._tensor_name, orig)
```

File: pytorch_lightning/callbacks/pruning.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/pruning.py:411]: Done !

File: pytorch_lightning/core/lightning.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:117]: Done !

File: pytorch_lightning/core/lightning.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:2017]: We should skip it as setting None would be override the base function if they exist. This would be wrong.

**Comment by tchaton** [comment]
> If I understand this PR correctly, all the changes are specific to the pruning etc. callbacks. can you confirm the following:
> 
> * pickling behavior unchanged
> * deepcopy(model) still does a real deepcopy unless the context manager is used
> * the context manager is not meant to be used by the user, or is it? any other use cases for deepcopying a lightnignmodule?

Yes, exactly !

**Review by awaelchli — APPROVED** [review]
Great!

Excluding the dataloaders here may not be necessary in the future if we avoid this weird patching of LightningModule methods: #6030

**Review by carmocca — APPROVED** [review]
